Add a stopper to DistSender

bdarnell opened the pull request:
This is primarily to prevent endless retries of asynchronous operations
like intent resolution if a retryable error occurs during shutdown.

Make multiTestContext's "store is stopped" errors retryable.
This fixes flakiness in TestSplitSnapshotRace_SnapshotWins.

Fixes #3121

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3898)

<!-- Reviewable:end -->

bdarnell committed:
Respect RPCRetryOptions.Closer in DistSender.

This is primarily to prevent endless retries of asychronous operations
like intent resolution if a retryable error occurs during shutdown.

bdarnell committed:
Make multiTestContext's "store is stopped" errors retryable.

This fixes flakiness in TestSplitSnapshotRace_SnapshotWins.

Fixes #3121

mrtracy commented on server/server.go:
```diff
@@ -146,6 +146,7 @@ func NewServer(ctx *Context, stopper *stop.Stopper) (*Server, error) {
 	retryOpts.Closer = stopper.ShouldDrain()
```
So if you notice the comment above this, I already added some support for this when creating the server.  However, it appears that the "Closer" option on retryOpts is being overridden explicitly in at least one place by the DistSender, and therefore this custom retryOpts seems to be unnecessary.

bdarnell replied:
Ah, right. So I guess what I should do instead of introducing the new field is make all the tests set RPCRetryOptions appropriately.

mrtracy commented:
LGTM

mrtracy commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898#-K86lkwLFrwrnkm8wAaa-r2-146:-K86z4frEu7ITcf6HZJ9:1641069046) ([raw file](https://github.com/cockroachdb/cockroach/blob/e00d0ecf7fdca6f76c731209b43df9dd62bc7a81/server/server.go#L146)):*
I'm actually fine either way; I was very close to putting to stopper on distsender myself, but  just set the Closer field of RetryOptions instead.  If you think its easier to understand as a top-level component of DistSender, you won't get much of an argument for me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898#-K87-uZ906lPaC1nTEa7:-K87-uZ906lPaC1nTEa8:-2117494217) ([raw file](https://github.com/cockroachdb/cockroach/blob/a42610f26e910ecc315cf50574b6b10a0e2b6bcc/kv/dist_sender.go#L726)):*
s/Fatalf/Fatal/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/dist_sender.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898#-K87-uZ906lPaC1nTEa7:-K87RHat957_WyZUoeni:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a42610f26e910ecc315cf50574b6b10a0e2b6bcc/kv/dist_sender.go#L726)):*
Done.

---

*[server/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898#-K86lkwLFrwrnkm8wAaa-r2-146:-K87RJNG47tiPjhOkCcf:219238501) ([raw file](https://github.com/cockroachdb/cockroach/blob/e00d0ecf7fdca6f76c731209b43df9dd62bc7a81/server/server.go#L146)):*
No, it doesn't make sense to add a new field and clobber RetryOptions.Closer with it when we could just pass it in that way to begin with. Updated all the call sites to do that and removed the new field.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3898)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
Evict range descriptor cache on NotLeaderError with unknown leader

This fixes the last remaining source of flakiness in
TestSplitSnapshotRace_SnapshotWins. (The test's DistSender had cached a
pre-split version of the range descriptor, so requests that should have
gone to the (live) right side range were instead getting sent to
the (dead) left side).

bdarnell commented:
Added a new commit here that fixes another source of instability in this test. It now seems to be stable, although this has made me nervous about DistSender's range descriptor cache. We may need some sort of fallback to help catch recurring errors that could be solved with a cache invalidation that are not caught on any of the current code paths.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -707,6 +721,13 @@ func (ds *DistSender) sendChunk(ctx context.Context, ba roachpb.BatchRequest) (*
+		} else if !finished {
+			select {
+			case <-ds.rpcRetryOptions.Closer:
+				return nil, roachpb.NewErrorf("shutting down"), false
```
we have a `NodeUnavailableError` somewhere. Not clear what the semantics should be though.

bdarnell replied:
Yeah, we're not doing much with `NodeUnavailableError` but it looks like the semantics we want. Changed.

tschottdorf commented:
LGTM. Agreed that I'm also less confident in the seamless operation of the various caches than I would want to be.
